sstables/writer.hh

```diff
@@ -41,14 +41,27 @@ class metadata_collector;
 class file_writer {
     output_stream<char> _out;
     writer_offset_tracker _offset;
+    std::optional<sstring> _filename;
+    bool _closed = false;
 public:
-    file_writer(file f, file_output_stream_options options)
-        : _out(make_file_output_stream(std::move(f), std::move(options))) {}
+    // It is the responsiblity of the caller to close file f
+    // if the constructor throws
```

Comment:
This suggests that the constructor is doing too much. Maybe the constructor should be private and there should be a public static make() function that can handle exceptions.

Reply:
@espindola why? the caller created f, so the caller is responsible for closing f, no?

Reply:
My preference is for constructors that are super simple (noexcept in particular). That way ownership transfer is simple, once the constructor is called the destructor is responsible for freeing resources.

Any convenience that is lost by the simple constructor (like being able to call non noexcept functinos like  make_file_output_stream) is regained by moving that to a static make function.

That is just a preference and should not block merging this change.

Reply:
@espindola so let's defer that and do it as a follow up as making the file output stream in the constructor is an existing problem.

I opened https://github.com/scylladb/scylla/issues/5901 for doing that.